### ISSUE TITLE
Show progress bar when fetching messages from mailserver

### ISSUE THREAD

**Issue by pedropombeiro** [status-im/status-react#4345]
Blocked by: https://github.com/status-im/status-go/issues/1036

### User Story

As a user, I want to know when the app is fetching messages from the mailserver so I'm not wondering if the app is doing anything, or if my push notification was bogus. 

cc @andmironov for design input. My suggestion would be to have an indeterminate progress indicator near the top of the mobile app, like this (the indicator pictured is not indeterminate, but you get the gist):

![image](https://user-images.githubusercontent.com/138074/40244503-5000107e-5ac3-11e8-9abc-567bae9ccb63.png)

### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Feature

[comment]: # (Describe the feature you would like, or briefly summarise the bug and what you did, what you expected to happen, and what actually happens. Sections below)
*Summary*: It is confusing to receive a push notification telling me I have a new message, clicking on it and then seeing no new messages. Since the app only fetches messages after being brought to the foreground, we should show an indeterminate progress bar to inform the user that the app is working on something.

#### Expected behavior
[comment]: # (Describe what you expected to happen.)

1. I receive a Push Notification
1. I tap the PN
1. The app opens on the main screen, and a progress bar indicates it is fetching messages
1. After a few seconds, new unread messages appear
1. After a predefined time limit or if a few more seconds have gone by without no new messages appearing, the progress bar is removed

#### Actual behavior
[comment]: # (Describe what actually happened.)

1. I receive a Push Notification
1. I tap the PN
1. The app opens on the main screen, and no new unread messages are visible
1. After a few seconds, new unread messages appear

### Solution
[comment]: # (Please summarise the solution and provide a task list on what needs to be fixed.)
*Summary*: 

Right now it is not possible to know whether messages which are arriving come from the mailserver, nor if/when they have finished arriving. Therefore we have to do a best effort solution for now. For instance, we could follow these heuristics:
- Keep progress bar visible if `shhext_requestMessages` was called less than 10 seconds ago
- Keep progress bar visible if a new unread message was added less than 5 seconds ago
- Hide progress bar if `shhext_requestMessages` was called more than 30 seconds ago

**Comment by status-open-bounty** [comment]
![Contract deploying](https://openbounty.status.im/img/deploying_contract.png)
Transaction: [0xe970c53aec33d45c1055dd40781ea0e8a6c7fbaa906c91df65dc201bf68b458a](https://etherscan.io/tx/0xe970c53aec33d45c1055dd40781ea0e8a6c7fbaa906c91df65dc201bf68b458a)

**Comment by hesterbruikman** [comment]
@PombeirP could it be this issue was resolved in the meantime through a grey notification bar (aka snackbar)?

Related issues:
https://github.com/status-im/status-react/issues/5371
https://github.com/status-im/design/issues/28

**Comment by pedropombeiro** [comment]
@hesterbruikman In my humble opinion, I think that what fits best in this scenario is a progress bar. Are you suggesting that we do a temporary fix to resolve https://github.com/status-im/status-react/issues/5371, or actually change this issue to be about implementing a snackbar?

**Comment by hesterbruikman** [comment]
I agree that a snackbar might not be the best fit here. As the snackbar was implemented I was curious if it had been in response to this issue (whether it's the best fit or not), or if no action had been taken on this issue yet.

I would say there's a temp fix opportunity to improve the snackbar and a long term fix opportunity to consider alternatives like progress bar, or pattern of notification. 

@errorists is involved in both issues; so if there's any way these issues can be unified into one solution that offers a good fit for both cases, I'm sure he'll come up with it.

(And welcome back:))

**Comment by pedropombeiro** [comment]
Thanks @hesterbruikman :-) I don't think anything was done regarding this issue (and I wasn't aware that we had changed to a snackbar), but the progress bar solution seems fairly nonintrusive and easy to implement, so hopefully we can move in that direction soon.

**Comment by errorists** [comment]
Here's a proposed design, similar to WhatsApp, Telegram and many more. I think we should include all 'connection' related messages to use this pattern instead of the snackbar which blocks the view of tappable items. 

https://framer.cloud/MeuuW/

<blockquote><img src="https://framer.cloud/MeuuW/6/framer/social-1200x630.png" width="48" align="right"><div><strong><a href="https://framer.cloud/MeuuW/">Status Status Banners</a></strong></div><div>This project was made with Framer, a design tool for turning original concepts into working prototypes.</div></blockquote>

**Comment by errorists** [comment]
And here's the design spec. 

<img width="497" alt="slice" src="https://user-images.githubusercontent.com/4932774/45941219-69e39400-bfdd-11e8-92c6-f3cba2f0ed0a.png">

I'd like us to use a custom spinner there, which file format would we want to use here
• Lottie https://airbnb.design/lottie/
• APNG
• GIF
• PNG sequence
?

A nice to have -> Animate displaying the Activity notification
Animated.timing(
value: opacity from 0 to 1 
duration: 100
easing: Easing.inOut
)
and reverse the animation when the notification is hidden


<blockquote><img src="http://airbnb.design/lottie/static/images/lottie.png" width="48" align="right"><div>Lottie</div><div><strong><a href="http://airbnb.design/lottie/">Lottie</a></strong></div><div>Easily add high-quality animation to any native app. Lottie is an iOS, Android, and React Native library that renders After Effects animations in real time, allowing apps to use animations as easily as they use static images.</div></blockquote>

**Comment by chadyj** [comment]
@errorists @EugeOrtiz what does this look like on desktop?

**Comment by errorists** [comment]
<img width="1024" alt="public chat - messages copy 3" src="https://user-images.githubusercontent.com/4932774/46952837-5a131900-d08c-11e8-8c46-4ca305c0d8d9.png">


I'd suggest to place it like so, I believe the left sidebar on desktop is the better choice for status notifications like this one. @EugeOrtiz wdyt?

**Comment by chadyj** [comment]
Got some mobile design feedback on this I wanted to share:

Hiding the username with the spinner isn't ideal because it blocks the user from knowing who they are chatting with. If "fetching" displays for some time (as it currently does) then the username is blocked for the duration.

I tend to agree.

**Comment by errorists** [comment]
Thanks for input! It's a tough one to answer because I thought about it while working on the design and did a couple of minor tweaks to help with any confusion, like displaying the notification with a slight delay after you enter a chat. In real-world use I personally doubt it will be much of an issue, since entering a chat with someone is a deliberate action and you're more than likely to remember with whom you wanted to chat or recognise the context — but I'd love to test it and be proven wrong. I'll give the design some more thought since I believe that overall it's a huge improvement, in that it doesn't cover any actionable - tappable items :)

**Comment by chadyj** [comment]
> since entering a chat with someone is a deliberate action

Sure, but what about returning to the phone after an hour, or the next day, or other use cases where it is no longer deliberate?  You could design around it with a flash/delay but that seems like working around the problems instead of fixing the cause.

On mobile with limited screen real estate the recipients name is crucial, so are there really no alternatives?

Testing sounds great but might be hard to test for this with mockups?

**Comment by chadyj** [comment]
@errorists  Got some other feedback about another point:

The chat screen is used to send and receive transactions, so it is important for a user to be certain who they are chatting with at all times.  Hiding the user could lead to accidental transactions.

I'd love to see other concepts before we commit to further work on this.

**Comment by errorists** [comment]
So another approach would be to make room for it, like lowering the entire content scroll area like pictured below. Not sure how feasible it is with React tho?

<img width="855" alt="screenshot 2018-10-15 at 16 47 00" src="https://user-images.githubusercontent.com/4932774/46958388-01e31380-d09a-11e8-912a-b60d1ac20cb8.png">

**Comment by EugeOrtiz** [comment]
@errorists we have more functionality coming for desktop, like _search_ that I think could be placed there. 
Why do we want to change the current badges? Is it only aesthetics? If we want to redesign this, I would look for something more temporal, that doesn't need the original UI real state. Also, on desktop, I imagine notifications that are general (placed in a general space) and notifications that are more specific to a certain chat/person and they should be placed accordingly (e.g. X person is offline).   **zpl.io/V1qZyPk**

Regarding the original issue, the fetching problem should be solved by now with current badges, right @PombeirP ?

![img_7109](https://user-images.githubusercontent.com/37297779/46960551-1970cb00-d09f-11e8-859a-7c20588501bf.PNG)

**Comment by errorists** [comment]
@EugeOrtiz because they [obstruct actual functionality](https://github.com/status-im/status-react/issues/5371). You can't add a contact till they're there, so my morning routine looks like this: I have a couple of new contacts and I wait and wait till that fetching messages dialogue is gone till I can add them.
But aesthetics too, they're too much in your face, being by far the most prominent element on your screen whenever I reopen the app. I'd too love to reduce it to a little temporal element that gets out of the way but then we're stripping away the message on what is actually happening in the background :)

**Comment by EugeOrtiz** [comment]
I see, we don't have much of that problem on desktop _I think_. Pushing the content for as you suggest could be a quick solution.

**Comment by pedropombeiro** [comment]
@errorists do you feel it is absolutely necessary to have the `Fetching messages` text message visible? Wouldn't the unobtrusive horizontal marquee progress bar be enough by signaling to the user that work is being done in the background, without revealing the exact nature of the work?

**Comment by pedropombeiro** [comment]
@errorists do you feel it is absolutely necessary to have the `Fetching messages` text message visible? Wouldn't the unobtrusive horizontal marquee progress bar be enough by signaling to the user that work is being done in the background, without revealing the exact nature of the work?

**Comment by errorists** [comment]
Hey, sorry for taking a while but here's the documentation for the fetching progress bar design.

[Link to Prototype ](https://framer.cloud/MsLgV/)
It's an indeterminate progress bar however, I'd suggest to also implement the determinate one when working on it so we can use the component in other areas like the [Browser](https://github.com/status-im/status-react/issues/6851) 🙏 

To retrace what's happening in the design in more details

1. The progress bar track is rendered directly above the top navigation bar, aligned to its bottom. It's 3px high with Light Blue background colour. 
Animation: value: opacity from 0 to 1, duration: 100ms, easing: easeOut
2. The progress bar indicator is placed on the left edge of the track, it's using our Blue accent colour, it's width is set a 0% of the track, height is 3px, border-radius is 1.5px.
3. Progress indicator animates to 
- X position : 20% of total track width
width: 60% of total track width
duration: 400ms, easing: easeIn
- X position : 85% of total track width
width: 15% of total track width
duration: 400ms, easing: easeOut
- X position : 20% of total track width
width: 60% of total track width
duration: 400ms, easing: easeIn
- X position : 0% of total track width
width: 15% of total track width
duration: 400ms, easing: easeOut

And it repeats after that :)

4. Progress bar can be terminated at any time, animates to: value: opacity from 1 to 0, duration: 100ms, easing: easeIn.

For a determinate progress bar, just animate the width of the track from 0% to 100%, easing: ease, duration: should be somehow proportionate to the distance animated so the entire width is 1 second, then animating from say 20% to 50% takes 300ms.

Please, let me know if further help is required, thanks!

**Comment by flexsurfer** [comment]
Implementation of indicator animation: https://github.com/status-im/status-react/pull/7000/files#diff-5a1ce6b28582016d18f9189701433442R107

<blockquote><img src="https://avatars3.githubusercontent.com/u/11790366?s=400&v=4" width="48" align="right"><div><img src="https://assets-cdn.github.com/favicon.ico" height="14"> GitHub</div><div><strong><a href="https://github.com/status-im/status-react/pull/7000">[#6851] Fix loading indicator in browser by flexsurfer · Pull Request #7000 · status-im/status-react</a></strong></div><div>fixes #6851</div></blockquote>

**Comment by yenda** [comment]
@errorists how long should the green "Connected" bar stay before disappearing? 
If you can also specify the animation like you did for loading that'd be awesome

**Comment by errorists** [comment]
Cool. It shouldn't stay long, `1500ms` is enough. The animation for the bar:
• Value: `Opacity from 0 to 1`, 
• easing: `ease`, 
• duration: `150ms`. 
Reverse on disappearing.

In the design it uses a scaleY but in React Native we can't set the Transform Origin property so it won't work like that. Alternative solution, if the snackbar is placed below the Top Bar in stack, we could animate it along the Y axis so it slides in from below it, the easing and duration would then be the same as above, the value would then be `Y position` from `minus snackbar's total height` to `0`.

If it's easy to add: changing the color of the snackbar should animate from 'Grey' to 'Green' with `easing: ease` and a `100ms duration`
